### PR TITLE
Change the formatter API to return an exception

### LINKED ISSUE #26381: [Formatter] Change the APIs to return an exception
**Description:**
When an exception occurs during formatting, the current implementation logs the exception and returns back an unaltered syntax tree. This behavior should be changed to return an exception so that the module using the formatter core can handle it accordingly.

**Steps to reproduce:**

**Affected Versions:**
Swan Lake Preview 4

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by IrushiL**
## Purpose

Fixes #26381

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/Formatter.java [hunk 1]
```diff
@@ -94,19 +96,20 @@ public static SyntaxTree format(SyntaxTree syntaxTree, LineRange range, Formatti
...
+    private static SyntaxTree modifyTree(SyntaxTree syntaxTree, FormattingOptions options, LineRange range)
+            throws FormatterException {
         FormattingTreeModifier treeModifier = new FormattingTreeModifier(options, range);
         ModulePartNode modulePartNode = syntaxTree.rootNode();
         try {
             return syntaxTree.modifyWith(treeModifier.transform(modulePartNode));
         } catch (Exception e) {
-            LOGGER.error(String.format("Error while formatting the source: %s", e.getMessage()));
-            return syntaxTree;
+            throw new FormatterException("Error while formatting the source: " + e.getMessage(), e.getCause());
```
Comment (nadeeshaan) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/Formatter.java:112]: Does `e.getMessage()` always return the source name? Otherwise let's change the exception message

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java [hunk 2]
```diff
@@ -59,8 +59,13 @@ public void test(String source, String sourcePath) throws IOException {
...
         TextDocument textDocument = TextDocuments.from(content);
         SyntaxTree syntaxTree = SyntaxTree.from(textDocument);
-        SyntaxTree newSyntaxTree = Formatter.format(syntaxTree);
-        Assert.assertEquals(newSyntaxTree.toSourceCode(), getSourceText(assertFilePath));
+        try {
+            SyntaxTree newSyntaxTree = Formatter.format(syntaxTree);
+            Assert.assertEquals(newSyntaxTree.toSourceCode(), getSourceText(assertFilePath));
+        } catch (FormatterException e) {
+            // TODO: handle the test cases causing formatting exceptions
+//            Assert.fail(e.getMessage(), e);
```
Comment (nadeeshaan) [inline misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java:67]: Are we going to add test cases for these particular negative cases or just ignore them? If ignoring let's remove the commented code

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java (hunk 2 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java:67]: These test cases should ideally be enabled since these are scenarios with no diagnostic issues (positive test cases).

But at the moment these are failing due to some parser issues which are yet to be fixed. Once they are fixed this can be enabled, and ideally, these test cases shouldn't fail.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/Formatter.java (hunk 1 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/Formatter.java]: Fixed

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java [hunk 3]
```diff
@@ -74,8 +79,13 @@ public void testParserResources(String sourcePath) throws IOException {
         TextDocument textDocument = TextDocuments.from(content);
         SyntaxTree syntaxTree = SyntaxTree.from(textDocument);
         if (!syntaxTree.hasDiagnostics()) {
-            SyntaxTree newSyntaxTree = Formatter.format(syntaxTree);
-            Assert.assertEquals(newSyntaxTree.toSourceCode(), getSourceText(filePath));
+            try {
+                SyntaxTree newSyntaxTree = Formatter.format(syntaxTree);
+                Assert.assertEquals(newSyntaxTree.toSourceCode(), getSourceText(filePath));
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java:84]: Can directly use `content` here

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java (hunk 2 above)
Comment (SupunS) [reply misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java:67]: I think its better to fail the test here and disable any-failing tests, if there are issues from the parser. 
Otherwise, any newly failing test would go un-noticed .

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java (hunk 3 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java]: Done

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java (hunk 2 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/FormatterTest.java]: Done

**Review by SupunS — APPROVED** [review]
